rate limite etcd healthcheck request

aojea opened the pull request:
/kind bug

#### What this PR does / why we need it:

Etcd healthcheck create one request to etcd per healthcheck requested, this can cause an excessive traffic going from apiserver to etcd.

Implement a rate limiter based on the configured timeout of the healthcheck, so all connections that are rate limited return the last known value.



```release-note
apiserver /healthz/etcd endpoint rate limits the number of forwarded health check requests to the etcd backends, answering with the last known state if the rate limit is exceeded. The rate limit is based on 1/2 of the timeout configured, with no burst allowed.
```

aojea commented:
/assign @deads2k @soltysh @tkashem 

I think that this is a better solution for the etcd healthzchecks, I'm not familiar with the code so I plumbed it here, but it can go directly into the etcd healthcheck

Looking for feedback first, I will add tests if this is an acceptable solution

/hold

aojea commented:
/sig api-machinery

aojea commented:
:thinking: :eyes: 

/test pull-kubernetes-integration

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:
```diff
@@ -279,8 +281,13 @@ func handleRootHealth(name string, firstTimeHealthy func(), checks ...HealthChec
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		err := c(r)
+		_, err, _ := requestGroup.Do(checkKey, func() (interface{}, error) {
+			err := c(r)
```
this is passing the first request to the health check. if it uses the request's context for cancellation, that could timeout / error the check for other requests with different timeouts

aojea replied:
Does the request to healthcheck goes through the timeout handlers?

liggitt replied:
yes, but I'm not sure why that's relevant

imagine:
1. request A to /healthz arrives with timeout of 1 second
2. request B to /healthz arrives with timeout of 30 seconds and gets aggregated into the singleflight
3. /healthz takes longer than a second
4. request A times out, failing the /healthz check with a timeout error
5. the timeout response is sent to request B

aojea replied:
do you mean something like `https://host:6443/healthz?timeout=5?`

AFAIK Unless you have a special handler , you don't have per request timeouts on the server https://github.com/golang/go/issues/54136 , or am I getting this wrong?

liggitt commented:
the etcd check specifically was attempted to be made non-blocking in https://github.com/kubernetes/kubernetes/pull/104437, but that didn't get completed

aojea commented:
/test pull-kubernetes-integration

aojea commented:
/test pull-kubernetes-integration

aojea commented:
> the etcd check specifically was attempted to be made non-blocking in #104437, but that didn't get completed

let's target the etcd check specifically, I've also found a bug in current logic

aojea commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -161,29 +162,36 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
-		defer cancel()
-		// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
-		_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
+		// Coalesce concurrent requests to avoid excessive traffic from the
+		// healthz endpoint to etcd
+		_, err, _ := requestGroup.Do(checkETCDKey, func() (interface{}, error) {
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
+			_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
+			return nil, err
+		})
```
@liggitt this a more targeted fix, only the calls that need to get the request from etcd will be coalesce, we avoid any assumptions on the http middleware of the `healthz` endpoints

I saw in the other issue you linked that you were thinking on limit rating this check ... I can implement that option too, I don't see it will be hard, add a rate limit and serve from the cache the ones that overflow ...

aojea replied:
I thought more about this and I think it is better to use the rate limiter instead,

aojea commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -161,29 +162,36 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		// shutdown for additional 2s seems acceptable.
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
```
@wojtek-t @liggitt I think this is a bug, since all healtcheck calls are serialized, this fixes overwhelm etcd, but it should be adding a considerable delay to clients

liggitt replied:
you're saying it was a bug before that would serialize health check calls, and is fixed in this PR, right?

aojea replied:
yes, every call to the healthcheck holds the lock, thus , all the calls has to be serialized

liggitt replied:
that was true previously, and is no longer true in this PR? just making sure I'm reading this correctly

aojea replied:
yes, I've added it as a separate commit to differntiate the changes

aojea committed:
fix etcd unit tests

stop leaking goroutines

reduce etcd test duration

aojea committed:
rate limit /healthz etcd healthchecks

return the last request error, instead of last error received
The rate limit allows 1 event per healthcheck timeout / 2

aojea commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory_test.go:
```diff
@@ -240,3 +242,123 @@ func TestCreateReadycheck(t *testing.T) {
 	}
 }
+
+func TestRateLimitHealthcheck(t *testing.T) {
```
```
$ $ go test -timeout 120s k8s.io/apiserver/pkg/storage/storagebackend/factory -v -race -run TestRateLimitHealthcheck -c
$ stress ./factory.test -test.run TestRateLimitHealthcheck
5s: 0 runs so far, 0 failures
10s: 24 runs so far, 0 failures
15s: 25 runs so far, 0 failures
20s: 48 runs so far, 0 failures
25s: 72 runs so far, 0 failures
30s: 72 runs so far, 0 failures
35s: 96 runs so far, 0 failures
40s: 116 runs so far, 0 failures
45s: 120 runs so far, 0 failures
50s: 144 runs so far, 0 failures
^C
```

aojea commented:
/assign @liggitt @lavalamp 

as original reviewers of https://github.com/kubernetes/kubernetes/pull/104437

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -161,33 +179,43 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
```
this is mutating variables, why is this not a write lock?

aojea replied:
IIUIC this lock is only for `clientErr`  ... but now that you mention it,  since we have the atomicError, we can use it now ,

aojea replied:
ahh, my bad, is locking the `client` too

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -123,11 +124,28 @@ func newETCD3ReadyCheck(c storagebackend.Config, stopCh <-chan struct{}) (func()
+}
+
+func (a *atomicError) Load() error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
```
RLock/RUnlock?

aojea replied:
yes

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -161,33 +179,43 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
+			err = fmt.Errorf("error getting data from etcd: %w", err)
 		}
-		return fmt.Errorf("error getting data from etcd: %w", err)
+		lastError.Store(err)
```
I'm trying to decide how big of a deal it is that a slow failed response to etcd can overwrite a fast successful response from etcd that started later:

1. client X hits /healthz, triggers request A to etcd
2. client Y hits /healthz, triggers request B to etcd
3. request B completes successfully, stores nil
4. request A completes unsuccessfully (maybe was to a dead etcd member, took 2 seconds, timed out, etc), stored error
5. client Z hits /healthz, hits rate limiter, gets cached error response B

Should we remember the time at which we started the `client.Get` request, pass it to the Store function, and make storing conditional on not overwriting a result that started later?

aojea replied:
I've added a test case, the use case is legit, but I think is going to be very rare because the rateLimiter depends on the healthzcheck timeout, the window is very small ... but definitively possible and is easy to cover it

aojea replied:
check the test  I've added, is easier to see there what  I mean

aojea committed:
don't serialize etcd healthchecks

deads2k commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -169,25 +191,37 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
+		if limiter.Allow() == false {
+			return lastError.Load()
+		}
 		ctx, cancel := context.WithTimeout(context.Background(), timeout)
```
is timeout guaranteed to be shorter than the timeout on the request?  Should it be?

liggitt replied:
hmm... request isn't even plumbed to this method, so... no, the particular request triggering the call to this check could be cancelled or timeout before this completes

aojea replied:
I have on my TODO to understand the request path on the /healthz endpoint, but the request context is independent of the healthcheck 

https://github.com/kubernetes/kubernetes/blob/97b58ece4a2ad1a366a9cf7deea9581373931238/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L205-L207

but it seems this was always this way with this healthcheck

liggitt replied:
That's true, but in this PR, one request's result can impact the results returned to other requests, so timeout/failure/cancellation is now relevant beyond just the single request

I want to make sure we don't have a weird interaction where:
* a request consumes a rate limiter token
* etcd health check gets spawned
* request cancels/terminates quickly
* timeout filter panics/interrupts the etcd health check in a way that keeps the cached result from being refreshed

aojea replied:
I see, I got it now, ... based on the handler code linked above,  the request lifecycle is independent of the healthcheck lifecycle ... the context passed to the etcd client probe is based on the configured healthcheckTimeout ... it seems safe

tkashem replied:
I think with the current implementation, all in-flight `readyz` and `healthz` requests will need to contend for the lock, so in essence all healthz/readyz probes are serialized, if i am not mistaken.

https://github.com/kubernetes/kubernetes/blob/6e3d62ca1c9e1132be5cd0be6746512fdd411b60/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L172-L191

if we want a constant lookup for the probes, and we want probe requests in flight to be independent of each other then making the actual check async as shown in https://github.com/kubernetes/kubernetes/pull/104437 is a safer option.

@aojea I have a draft https://github.com/kubernetes/kubernetes/pull/112110 for reference

aojea replied:
Do we want to create a periodic probe?

aojea replied:
that is simpler than this, of course

aojea replied:
> I think Jordan asked the same question here , #112046 (comment)
> what timeout is the one we are taking about here? The apiserver handler has a special handler that derives the timeout form the args of the request, but I think it does not apply here
> 

@liggitt @deads2k I've asked @sttts about the handler and I was wrong, timeout on requests are respected on this handler too , I've confirmed it:

> kubectl get --raw https://1/healthz/etcd?timeout=1us
> Error from server (Timeout): the server was unable to return a response in the time allotted, but may still be processing the request

The answer should be cached though

lavalamp replied:
yeah it seems correct that this check continues even if the triggering request is cancelled/timed out.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go:
```diff
@@ -123,20 +124,42 @@ func newETCD3ReadyCheck(c storagebackend.Config, stopCh <-chan struct{}) (func()
+	return a.err
+}
+
 func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan struct{}) (func() error, error) {
```
Do we construct one of these for every resource? Or one per backend?

aojea replied:
hmmmm https://github.com/kubernetes/kubernetes/pull/111559 :/

aojea replied:
the stack is

```
newETCD3Check
-- newETCD3ReadyCheck
-- newETCD3HealthCheck
---- CreateHealthCheck
---- CreateReadyCheck
------- addEtcdHealthEndpoint
-----------ApplyWithStorageFactoryTo
```

if I test sending multiple requests to the endpoints, doing this 

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
index 55e2cc0b42c..908f8717957 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
@@ -168,6 +168,7 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
                        clientErr = fmt.Errorf("server is shutting down")
                }
        }()
+       counter := 0
 
        return func() error {
                // Given that client is closed on shutdown we hold the lock for
@@ -180,6 +181,8 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
                if clientErr != nil {
                        return clientErr
                }
+               counter++
+               fmt.Println("DEBUG ETCD CHECK", counter)
                ctx, cancel := context.WithTimeout(context.Background(), timeout)
                defer cancel()
                // See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
diff --git a/test/integration/apiserver/apiserver_test.go b/test/integration/apiserver/apiserver_test.go
index c371ead248f..9215f002262 100644
--- a/test/integration/apiserver/apiserver_test.go
+++ b/test/integration/apiserver/apiserver_test.go
@@ -210,6 +210,29 @@ func Test4xxStatusCodeInvalidPatch(t *testing.T) {
        }
 }
 
+func TestEtcdHealthz(t *testing.T) {
+       server := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+       defer server.TearDownFn()
+
+       rt, err := restclient.TransportFor(server.ClientConfig)
+       if err != nil {
+               t.Fatal(err)
+       }
+
+       for i := 0; i < 100; i++ {
+               req, err := http.NewRequest("GET", server.ClientConfig.Host+"/healthz/etcd", nil)
+               if err != nil {
+                       t.Fatal(err)
+               }
+               resp, err := rt.RoundTrip(req)
+               if err != nil {
+                       t.Fatal(err)
+               }
+               resp.Body.Close()
+
+       }
+}
```

the output only shows one "counter"

```
DEBUG ETCD CHECK 62
DEBUG ETCD CHECK 63
DEBUG ETCD CHECK 64
DEBUG ETCD CHECK 65
DEBUG ETCD CHECK 66
DEBUG ETCD CHECK 67
DEBUG ETCD CHECK 68
DEBUG ETCD CHECK 69
DEBUG ETCD CHECK 70
DEBUG ETCD CHECK 71
DEBUG ETCD CHECK 72
DEBUG ETCD CHECK 73
```

it seems one , I don't think that this does per backend, that seems to be done inside the client. that has the endpoints of the etcd backends

lavalamp replied:
That seems pretty promising, an even more conclusive check would be counting how many times newETCD3Check is called, maybe?

lavalamp replied:
I'm not totally convinced the testing and real startup sequences are sufficiently similar about this though 😬

aojea replied:
I can run in a real clusters though, so we are confident

aojea replied:
verified

```
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
index 4c8a409d659..ed10ca7b513 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
@@ -41,6 +41,7 @@ func Create(c storagebackend.ConfigForResource, newFunc func() runtime.Object) (
 
 // CreateHealthCheck creates a healthcheck function based on given config.
 func CreateHealthCheck(c storagebackend.Config, stopCh <-chan struct{}) (func() error, error) {
+       fmt.Println("CreateHealthCheck")
        switch c.Type {
```


```
kubectl -n kube-system logs kube-apiserver-test-control-plane | grep CreateHealthCheck | wc
      1       1      18
```

lavalamp commented:
I think this is ready to go as long as we don't make N separate checkers...

leilajal commented:
/triage accepted

aojea commented:
> I think this is ready to go as long as we don't make N separate checkers...

kindly ping @lavalamp https://github.com/kubernetes/kubernetes/pull/112046#discussion_r960968248

lavalamp commented:
Sorry I missed your last update.

/lgtm
/approve

lavalamp commented:
If you can figure out an easy way to count how many health checks this does in e.g. our e2e tests before/after this change, and the difference is very large, a backport might be reasonable.

(Do we scrape etcd metrics at the test end? That would be the first place I'd look. This might not be super easy to measure.)

aojea commented:
> (Do we scrape etcd metrics at the test end? That would be the first place I'd look. This might not be super easy to measure.)

Mike Spreitzer added tooling to dump the metrics of etcd on the integration tests (all test share the same etcd instance)
You just download the tarball with the mtrics
```
wget https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/112252/pull-kubernetes-integration/1568270000328806400/artifacts/etcd-scrapes.tgz
```

and run the script he created pointing to the file

```
$HOME/go/src/k8s.io/kubernetes/hack/run-prometheus-on-etcd-scrapes.sh etcd-scrapes.tgz
```

I didn't find any metrics that shows the number of `Get` received on the healthz endpoint, or something that can give us an idea ... I've compared a dumped version from before and other from after this patch and I couldn't find any significant difference-


I don't think that our e2e abuse the /healthz/etcd endpoint , it seems Openshift does much more stress on the endpoint and that's why we were carrying over with David's patch

lavalamp commented:
OK, thanks for checking!

wojtek-t commented:
I missed that originally - but this looks great - I'm really happy that you made that happen.

Re @lavalamp questions - i don't think we would find anything in our tests (which don't overload any of the endpoints FWICT), but it certainly can help protecting some real-world scenarios.
But I don't think I can come up with a justifying argument for backport.
